components.md

```diff
@@ -0,0 +1,130 @@
+
+## Actor
+The actor participates in the network listening for commands and acts on them. 
+It may act regularly of its own accord.
+Typically an actor controls one or several devices via Drivers and/or does some kind of processing.
+Actions which the actor can take could be anything from measuring on a regular basis ("acting" as a sensor), commanding hardware devices through the use of the driver to do X or Y, processing some data and re-emitting results (e.g. a PID controller).
+
+Potentially a GUI could be attached here too.
+
+The Actor would send commands to a Driver, instructing it to `set`/`get`/`call` a given Parameter/Action by name, with a value if appropriate, and receive replies.
+
+This could look for example like this:
+:::{mermaid}
+sequenceDiagram
+    par
+        Actor1->>Driver42: GET temp_K
+        Driver42->>HW-on-COM1: "*TEMP?\n"
+        HW-on-COM1->>Driver42: "275.14\n"
+        Driver42->>Actor1: 275.14
+    and
+        Actor1->>Driver11: GET temperature
+        Note over Driver11: Fetched a fresh value 23ms ago, using cached value
+        Driver11->>Actor1: 300.1
+    end
+    Note over Actor1: computes average temperature, converts to degC
+    Note over Actor1: stores/caches value maybe
+    Actor1->>Observer1: "Average oven temperature" 14.47 degC
+:::
+
+:::{admonition} TODO
+We could achieve more complete separation and make the Actor purely about processing _inputs_, doing some _operation_ and generating _outputs_.
+This could be either stateless (temperature conversion) or stateful (PID control).
+Maybe a better name would be "Processor" then.
+This means that _all_ device-related stuff stays compartmentalized in the Driver, and an Actor would address that Driver over the network (to fetch the Parameter values it is interested in).
+
+Thoughts?
+:::
+
+## Driver
+A Driver is a component that interfaces with a (hardware) Device in a manner we don't specify, and that has a specific API on the ECP side.
+A driver must contain some object dealing with hardware communication.
+This may be a `pymeasure.Instrument` instance or something from another instrument library.
+This is the place where all instrument libraries (including pymeasure) wire their hardware interface classes into ECP.
+
+Concerning the ECP, we draw the abstraction boundary at the Driver -- the details on how this communicates with a Device (SCPI, dll, ...) should not be relevant for the protocol details.
+What we define is how the other ECP components interact with the Driver, how it determines and announces its capabilities, etc.
+
+A driver must contain/manage
+* An object that communicates with the connected Device, including managing its lifecycle (init, operations, shutdown)
+* The name of the connected instrument
+* A list of available Parameters (properties/attributes to get/set) and Actions(methods to call)
+* `set`/`get`/`get_all`/`call` interfaces (for incoming commands to use to act on Parameters and Actions)
+
```

Comment:
"I see now what you did there" - so I....well. What is described here now as Driver, is what I had in mind as Actor, with the Actor being allowed to do a bit more stuff. I guess it may make sense to separate the two further, where it then also starts making more sense to rename the Actor to _Processor_. For me the Driver actually was the `pymeasure.Instrument`, so not a newly defined/specified component of the ECP in itself, but whatever hardware communication object that may exist (already, or at least separate from the ECP). 

In that sense, I think we could add a _Processor_ component which is then more what is described above as Actor, while the Driver here is renamed as Actor, while the "some object dealing with hardware communication" remains becomes the Driver, outside of the ECP. However, as a different thing, we could leave it as bilderbuchi now suggested, and rename the Actor to Processor, without giving the "object dealing with hardware communication" a separate name for our discussion (or at least choose a different one to "Driver"). 

In general, how it is described here, sounds to me, when going bottom-up and ending at the Driver, exactly as yaq, the Driver being a yaq daemon. In yaq (or at least in the python examples I saw, I did not look at all implementations, in both in python and otherwise), the Driver communicates directly with the Device, instantiating e.g. a serial connection, having methods defined which deal exclusively with hwardware (HW) -comms and HW-control, and exposing those to clients. What was important to me, was to separate this ECP communication layer from the HW communication layer, which might be e.g. a pymeasure.Instrument. This would clearly separate what `pymeasure` (and many homebrew hardware drivers I have seen) does from what ECP should do - however, I realise that this might "not be up to the ECP" to decide, whether people who implement an ECP Driver (in the terms described in the currently proposed version) do "play nice" and separate it, or not. 
If the two are separated, and things stay that much modular, it becomes easy to re-use these software parts in other frameworks, in case that is wanted/necessary. We might want to encourage possible adopters to stay with this nice new thing we defined/built, and trying to lock them in is not the way I would like to go about it. Most likely they will see it in advance and not start at all - we should convince possible adopters with the content itself, so that they stay because they like it, not because it is a hassle to move their definitions to a different type of framework.

Reply:
Thank you for your thoughts, I have to reply later.

Reply:
I'm not sure where you seem to(?) get the idea from that I want to "lock in" people into this framework somehow -- that is not my intention at all! I'd like to fix whatever gave that impression.

On the contrary, I'm aiming for a design where you can use those other objects _unchanged_, and all you have to maintain is an interface class (the `Driver` subclass) that wires things correctly between a `Driver.list_parameters()` and however `self.interface` yields that information, or what `Driver.get_param(name)` needs to do to get this from `self.interface` (in case of pymeasure.Instrument, this might a simple `return self.interface.getattr(name)`.

Reply:
> What was important to me, was to separate this ECP communication layer from the HW communication layer, which might be e.g. a pymeasure.Instrument. 

Does what I sketched in my comments fulfill this? At some point, you will need interface between ECP and HW, and I intended (misunderstanding the intention behind the original discussion comments -- sorry!) the Driver to be that place:
* ECP messages going in/out
* internally mapping the messages to what the contained HW communication layer/object understands
* the HW communication layer responds, that reply gets transformed into an appropriate message

Reply:
We definitely should talk about that part (driver actor, hardware...) in an issue, where the discussion is more easy to find.

I thought as well, that a Driver is a pymeasure instrument (or similar) and the actor is the interface from the network to the driver.

Reply:
I like it. Can be resolved from my side

Reply:
I re-read @bklebel's initial comment here, and I think it's resolved the way he wanted to see.
The only point I'd still like to clarify is where he perceived "lock in" potential -- I want very much to avoid that!

Reply:
Thanks for all your work, I was indisposed the last weeks, sorry for the late reply. @bilderbuchi is correct, how you changed it now is actually quite what I had in mind, so this is resolved for me. The "lock in" potential I saw, was if Actors and Drivers are not separated. Because, if I have something like that (typical in the yaq daemons I saw, and a the Tango python code I looked at), it is quite difficult to take only the Driver part, and use it somewhere else (i.e. if I find a Tango device server for a specific instrument, it does not help me at all, as long as I do not set up the whole Tango thing in my setup). 
If I search for a framework to work with, and I find one which tells me to first write a proper Driver class, and then this will be taken on a piggyback-ride by an Actor, I personally feel safer about the time I might invest into the Driver-development, because if I discovered that I don't like how the Actor carries my Driver around, I can just pluck the Driver out, and put it into something else. If Driver and Actor are fused together as a standard, this makes the effort to separate the two typically so big that it's quicker to just grab the Device's manual and write a new Driver from scratch, i.e. this produces a high threshold to change. 
How it is written now is fine with me.

Reply:
Ah I see what you mean! Since the whole thing is meant to interface/consume other libraries, I don't see much of a danger of that happening.